omnigibson/sensors/vision_sensor.py

```diff
@@ -215,38 +208,75 @@ def initialize_sensors(self, names):
         """
-        # Standardize the input and grab the intersection with all possible raw sensors
-        names = set([names]) if isinstance(names, str) else set(names)
-        names = names.intersection(set(self._RAW_SENSOR_TYPES.keys()))
-
-        # Initialize sensors
-        sensors = []
+        names = {names} if isinstance(names, str) else set(names)
         for name in names:
-            sensors.append(lazy.omni.syntheticdata.sensors.create_or_retrieve_sensor(self._viewport.viewport_api, self._RAW_SENSOR_TYPES[name]))
-
-        # Suppress syntheticdata warning here because we know the first render is invalid
-        with suppress_omni_log(channels=["omni.syntheticdata.plugin"]):
-            render()
-        render()    # Extra frame required to prevent access violation error
+            self._add_modality_to_backend(modality=name)
 
     def _get_obs(self):
         # Make sure we're initialized
         assert self.initialized, "Cannot grab vision observations without first initializing this VisionSensor!"
 
         # Run super first to grab any upstream obs
-        obs = super()._get_obs()
+        obs, info = super()._get_obs()
+
+        for modality in self._modalities:
+            raw_obs = self._annotators[modality].get_data()
+            # Obs is either a dictionary of {"data":, ..., "info": ...} or a direct array
+            obs[modality] = raw_obs["data"] if isinstance(raw_obs, dict) else raw_obs
+            if modality == "seg_semantic":
+                id_to_labels = raw_obs['info']['idToLabels']
+                obs[modality], corrected_id_to_labels = self._remap_semantic_segmentation(obs[modality], id_to_labels)
+                info[modality] = corrected_id_to_labels
+            elif modality == "seg_instance":
+                id_to_labels = raw_obs['info']['idToLabels']
+                info[modality] = id_to_labels
+        return obs, info
+    
+    def _remap_semantic_segmentation(self, img, id_to_labels):
+        """
+        Remap the semantic segmentation image to the class IDs defined in CLASS_NAME_TO_CLASS_ID.
+        Also, correct the id_to_labels input with the labels from CLASS_NAME_TO_CLASS_ID and return it.
+        """
+        # Convert string IDs to integers
+        int_ids = set(int(id) for id in id_to_labels.keys())
+
+        # Determine if there are any new IDs that were not previously known
+        new_ids = int_ids - self._known_semantic_ids
+
+        # If there are new IDs, update _known_semantic_ids set and rebuild the key array
+        if new_ids:
+            self._known_semantic_ids.update(new_ids)
+            max_id = max(self._known_semantic_ids)
+
+            # Initialize the key array with a default value for unmapped IDs
+            key_array = np.full(max_id + 1, -1, dtype=np.int32)
+
+            # Populate the key array with the new IDs based on class name mappings
+            for int_id in self._known_semantic_ids:
```

Comment:
this is great! One more thing -- I think `self._known_semantic_ids` (and any other corresponding variables) are global in scope (i.e.: their info is uniform across all VisionSensors), and so I think that this information can be "shared" across all VisionSensors.

Can we make this a class variable? And have it cleared when VisionSensor.clear() is called. This way the amortization benefits can be leveraged over all VisionSensors, rather than per-vision sensor!

Reply:
Of course, that makes a lot of sense! Both `known_semantic_ids` and `key_array` are now class variables.